Compromised APCs now flicker their darkblue screen, and it cannot be seen from cameras.

zxaber opened the pull request:
## About The Pull Request
- APCs now show the dark blue effect periodically, once every six to ten seconds.
- This flicker is not visible from cameras, except if the mob is anyway within view.
- Malf AIs will see their own APCs flicker through cameras, but other AIs will not see them, except when the AI mob itself is nearby. 
## Why It's Good For The Game
This is one of those "sacrificing soul for gameplay" things.

Malf AI does not win against an aware and uninjured crew. That's just the nature of the game. It's not even a competition,  humans have too many tools that silicons simply have no answer for. That's fine, it makes the fight asymmetric (which is cool), and plays further into Malf AIs (and AIs in general) using underhanded tactics to quietly unravel their disadvantages. It makes stealth the greatest asset for an AI.

Which is why the current hacked APC sprites are not great. The AI, by its own nature, needs to be able to make the first move. But the game is full of tools that allow the crew to, either accidentally or on purpose, seek out the telltale sign of a Malf AI. It's poor design that the AI is encouraged to not use any of it's antag tools because even the process of preparing to buy them can place them in a losing position.
## Changelog
:cl:
balance: Emagged and Hacked APCs now occasionally flicker a blue effect, and the effect is not visible with cameras.
/:cl:

zxaber committed:
does the thing

zxaber committed:
These should be removed actually

MrMelbert commented:
I don't like this in its current state, but I had a discussion over zxaber on discord to make some tweaks to it.

YakumoChen commented:
hacking APCs is kind of a risk/reward. Malf's offensive tools are _very_ powerful. Overload can easily crit in one go, malf can remove all the atmos protections to plasmaflood with ease, autoborger, mech hacking, etc. That doesn't even get into AI being a powerhouse just with its default abilities, on top of borgs existing.

While being outed early is lame, it's also not really too hard, especially with stealthy APCs, to suddenly be throwing the switch on the nuke out of nowhere, 40 minutes in, with the fortified Door Core:tm:, while you also plasmaflood with all the fire alarms disabled.

moocowswag commented:
It looks really good but

> Malf AI does not win against an aware and uninjured crew. That's just the nature of the game. It's not even a competition, humans have too many tools that silicons simply have no answer for. That's fine, it makes the fight asymmetric (which is cool), and plays further into Malf AIs (and AIs in general) using underhanded tactics to quietly unravel their disadvantages. It makes stealth the greatest asset for an AI._

This isnt entirely true as long as malf ai's have borgs.

Malf ai is basically the most natural form of biddle tots / heretics. Competent silicons tend to have a laundry list of tasks that need to get done to improve how powerful their faction is before going loud, the placement of tesla coils/the prepping of the powernet and distro, the construction of traps and gauntlets, countering or denying special equipment from being produced. The only thing a malf ai doesnt truely have answers to is other antags that get free abilities out of nowhere and bumrush the sat as soon as they spawn.


The smallest stealth buffs can and will go a long way to ensuring that silicons have more time to actually do all of this.

GuillaumePrata commented:
AI's stealth gameplay being like "I hope no one walks into this room" or the worst version of "lets build 20 apcs in a room I'm sure no one will walk in" are just badly designed stealth gameplay and I'm glad this PR nukes a good slice of it.

Would be nice to have more reasons for the crew to interact with APCs, both antags and not, so we have more plausable deniability and paranoia around the station when a doctor finds a hacked APC in medbay instead of it just being an instant "Silicon valids" call.
Hopefully a future design angle we could explore.

And just to ideasguy, the soul APC icon could be used for when the AI triggers the doomsday device, having every APC go blue in the middle of it would be a nice mood to get everyone on a last push against the AI.

Yttriums commented:
No comment on the stealh gameplay aspects, but thematically it would be cool if the interface was malfunctioning or glitched rather than just being syndicate colored.

SomeoneYouProbablyKnow commented:
Half of malf gameplay is gaslighting the crew into focusing on bigger threats to the station. 
This will only force people to check APCs.

Hacked APC just needs more plausible denyability. But designing that is harder.

Pickle-Coding commented:
Could make emagged APCs have this UI instead of blue screen as well, just so people won't be absolutely certain it would be AI.

SkeletalElite commented:
> Would be nice to have more reasons for the crew to interact with APCs, both antags and not, so we have more plausable deniability and paranoia around the station when a doctor finds a hacked APC in medbay instead of it just being an instant "Silicon valids" call.

its still a silicons valid call because an emagged APC has its access restrictions completely removed where as one that is hacked by an AI has its access restrictions set so no one can access it

zxaber committed:
Change in scope

zxaber commented:
The nature of this PR has shifted away from the interface and towards showing the dark blue overlay occasionally. Some tricks were added that prevent it from being seen over cameras (except when the mob is in view of the APC anyway). This also means it's safe for a malf AI to take a holopad call in their core, as the calling mob will not see the APC flicker to the hacked overlay.

YakumoChen commented:
Can hacked APCs just be flatly Malf Blue for the AI that controls them? I don't want to glance at a camera to see if I've hacked it and not know unless I wait or click it

MrMelbert commented on code/modules/power/apc/apc_appearance.dm:
```diff
@@ -116,3 +106,19 @@
+	for(var/mob/viewer in view(src))
+		if(viewer.client)
+			mobs_to_show += viewer.client
+		if(istype(viewer, /mob/camera/ai_eye))
+			var/mob/camera/ai_eye/viewer_eye = viewer
+			if(viewer_eye.ai == malfai && viewer_eye.ai.client)
+				mobs_to_show |= viewer_eye.ai.client
```
If the sole intent of this code is to ensure ONLY the ai that has hacked it can see the animatoin, it's not necessary / the next block of code does this
```suggestion
```

zxaber replied:
This code, and the following, have been removed in favor of client_image_holder objects for the AI.

MrMelbert commented on code/modules/power/apc/apc_appearance.dm:
```diff
@@ -116,3 +106,19 @@
+/obj/machinery/power/apc/proc/flicker_hacked_icon()
+	var/image/hacker_image = image(icon = 'icons/obj/machines/wallmounts.dmi', loc = src, icon_state = "apcemag", layer = FLOAT_LAYER)
+	var/list/mobs_to_show = list()
+	for(var/mob/viewer in view(src))
```
```suggestion
	// Collecting mobs the APC can see for this animation, rather than mobs that can see the APC. Important distinction, intended such that mobs on camera / with XRAY cannot see the flicker. 
	for(var/mob/viewer in view(src))
```

zxaber committed:
Update code/modules/power/apc/apc_appearance.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

zxaber committed:
AI indicators

zxaber commented:
> Can hacked APCs just be flatly Malf Blue for the AI (and its borgs) that controls them? I don't want to glance at a camera to see if I've hacked it and not know unless I wait or click it

AIs now see their own hacked APCs using the same methods that hallucinations use. This works across the map, and will continuously show the dark blue panel to the owning AI exclusively.

Borgs don't get the vision, because there would be an excessive amount of bookkeeping to handle it for not really a lot of gain. Borgs can still see the flicker effect, like any other nearby mob.

YakumoChen commented:
Good enough for me 👍

zxaber committed:
Okay, this name is actually dumb

also reverting part of the APC interface change I meant to fully remove earlier.

zxaber committed:
makes the AI's indicator a hud element

zxaber commented:
Changed the AI's indicator to be a hud, rather than the object effect. This made me think about making it more hud-themed too, rather than just using the same bluescreen overlay.

![image](https://github.com/tgstation/tgstation/assets/37497534/b2b9bb34-17e7-4f29-8ba8-db6c08a58168)

Malf AIs can now see when their APCs flicker, just like mobs within range, but also get a permanent-shown red bracket indicator for hacked APCs. The bracket gives an at-a-glance view that the APCs is hacked, while the flicker can be used to reasonably know if a crew member passing the APC could have caught on.

zxaber committed:
barsign

zxaber committed:
Merge branch 'master' into chasin'_away_the_blues

MrMelbert commented on code/game/data_huds.dm:
```diff
@@ -554,3 +557,13 @@ Diagnostic HUDs!
 	holder.icon_state = "electrified"
 	set_hud_image_active(DIAG_AIRLOCK_HUD)
+
+/*~~~~~~~~~~~~~~~~~
+	Hacked APCs!
+~~~~~~~~~~~~~~~~~~*/
```
Don't copy this comment, all these headers do is mess up auto doc
```suggestion
/// Applies hacked overlay for malf AIs
```

MrMelbert commented on code/game/data_huds.dm:
```diff
@@ -554,3 +557,13 @@ Diagnostic HUDs!
+/obj/machinery/power/apc/proc/set_hacked_hud()
+	var/image/holder = hud_list[MALF_APC_HUD]
+	holder.loc = get_turf(src)
+	holder.plane = HUD_PLANE
```
I think you're supposed to use one of the `SET_PLANE`s for this, but also `HUD_PLANE` is for screen elements like inventory, so I don't think it's supposed to be used for this

MrMelbert commented on code/modules/power/apc/apc_main.dm:
```diff
@@ -482,6 +487,11 @@
 		force_update = TRUE
 		return
 
+	if((obj_flags & EMAGGED || malfai) && icon_state == "apc0") //apc0 means we're not broken and the cover's not open.
```
Checking icon state directly for code comparisons is very.. iffy

Can we not just check for broken or opened? In fact, don't we already do that a bit further up the proc?

zxaber replied:
My reasoning for it was more that we *only* want the overlay to show if the `apc0` sprite is showing, because no matter what, if the screen isn't visible, the overlay shouldn't be either.

That said, you're right, there are already checks for broken and opened earlier, so the check here doesn't mean much.

MrMelbert commented on code/modules/power/apc/apc_appearance.dm:
```diff
@@ -116,3 +106,16 @@
+		if(viewer.client)
+			mobs_to_show += viewer.client
+	if(malfai?.client)
+		mobs_to_show += malfai.client
```
```suggestion
		mobs_to_show |= malfai.client
```
Just in case

zxaber committed:
updates

Ghommie commented on code/modules/power/apc/apc_appearance.dm:
```diff
@@ -116,3 +106,16 @@
+	if(malfai?.client)
+		mobs_to_show |= malfai.client
+	flick_overlay_global(hacker_image, mobs_to_show, 1 SECONDS)
+	hacked_flicker_counter = pick(3, 4, 5) //The counter is decrimented in the process() proc, which runs every two seconds.
```
```suggestion
	hacked_flicker_counter = rand(3, 5) //The counter is decrimented in the process() proc, which runs every two seconds.
```

zxaber committed:
Update code/modules/power/apc/apc_appearance.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

zxaber committed:
Merge branch 'master' into chasin'_away_the_blues

zxaber committed:
merge conflict

zxaber committed:
Merge branch 'chasin'_away_the_blues' of https://github.com/zxaber/tgstation into chasin'_away_the_blues
